### PR TITLE
[wal] small fixes in SearchEndHeight & replay logic

### PR DESCRIPTION
**Pull request by melekes**
I was not able to reproduce #1600, but I did fix one possible panic and improved `SearchEndHeight` behaviour (see commit messages for details).

### CODE REVIEW

File: consensus/wal.go [hunk 1]
```diff
@@ -166,17 +167,25 @@ func (wal *baseWAL) SearchForEndHeight(height int64, options *WALSearchOptions)
 		for {
 			msg, err = dec.Decode()
 			if err == io.EOF {
+				// OPTIMISATION: no need to look for height in older files if we've seen height - 1
+				if lastHeightFound == height-1 {
```
Comment (ebuchman) [inline consensus/wal.go]: what if theres multiple EndHeight in a single file ? Shouldn't we just check `lastHeightFound > 0 && lastHeightFound < height` ?

File: consensus/wal.go [hunk 2]
```diff
@@ -166,17 +167,25 @@ func (wal *baseWAL) SearchForEndHeight(height int64, options *WALSearchOptions)
...
 				// do nothing
+				continue
 			} else if err != nil {
 				gr.Close()
 				return nil, false, err
 			}
 
 			if m, ok := msg.Msg.(EndHeightMessage); ok {
+				lastHeightFound = m.Height
 				if m.Height == height { // found
```
Comment (ebuchman) [inline consensus/wal.go]: better yet why not just exit on the else condition here ?

Comment (melekes) [reply consensus/wal.go]: hm.. but we need to read until the end to ensure there is no `height`

Comment (melekes) [reply consensus/wal.go]: if we just check else if `m.Height < height` and return, we might return false when if fact height exists...

Comment (ebuchman) [reply consensus/wal.go]: but we start from the top of the file right ? and the WAL already guarantees the height is increasing. so if we find a height, and its less than what we're looking for, we should be done. no ?

Comment (melekes) [reply consensus/wal.go]: 1 -> 2 -> 3

we're looking for 3. if we return on 1 or 2, it would be incorrect since there is 3

Comment (ebuchman) [reply consensus/wal.go]: yeh duh. larger heights are at the bottom of the file.

**Comment by ebuchman** [comment]
test fail ?!

**Comment by melekes** [comment]
```
proto: no slice oenc for *reflect.rtype = []*reflect.rtype
proto: no encoder for Tags []common.KVPair [GetProperties]
proto: no coders for common.KI64Pair
proto: no encoder for Fee common.KI64Pair [GetProperties]
proto: no slice oenc for *reflect.rtype = []*reflect.rtype
proto: no encoder for Tags []common.KVPair [GetProperties]
proto: no coders for common.KI64Pair
proto: no encoder for Fee common.KI64Pair [GetProperties]
```

what's up with ^^^?

**Comment by melekes** [comment]
```
I[05-24|11:55:37.229] Remove address from book                     pex=0 addr=022ec801d79025caab3afbbf816d92ff8450d040@127.0.0.2:6593 ID=022ec801d79025caab3afbbf816d92ff8450d040
I[05-24|11:55:37.229] Add our address to book                      pex=0 addr=022ec801d79025caab3afbbf816d92ff8450d040@127.0.0.2:6593
E[05-24|11:55:37.229] Dialing failed                               pex=0 addr=022ec801d79025caab3afbbf816d92ff8450d040@127.0.0.2:6593 err="Connect to self: <nil>" attempts=0
I[05-24|11:55:37.229] Add our address to book                      pex=0 addr=022ec801d79025caab3afbbf816d92ff8450d040@127.0.0.2:6593
```

```
E[05-24|11:55:46.824] Failed to save AddrBook to file              pex=2 file=/tmp/pex_reactor002376369/addrbook2.json err="open /tmp/pex_reactor002376369/write-file-atomic-fZNA0VOBq0oSNl7LEys03sXKcC0nkMCE: no such file or directory"
```

```
D[05-24|11:55:36.978] Received message                             pex=0 src="Peer{MConn{127.0.0.3:12787} cba813cebcd8775de8e167ce34d1f6968591d524 out}" chId=0 msg="[pexAddrs [022ec801d79025caab3afbbf816d92ff8450d040@127.0.0.2:6593]]"
E[05-24|11:55:36.978] Stopping peer for error                      pex=0 peer="Peer{MConn{127.0.0.3:12787} cba813cebcd8775de8e167ce34d1f6968591d524 out}" err="Error{`Received unsolicited pexAddrsMessage`}"
```

**Comment by ebuchman** [comment]
> anybody knows what's this about?

No. It's been bugging me too. I think it's because the proto files are from another repo ... I'd really like to eliminate that and move the KVPair into the ABCI proto spec. Not sure if we really need them in the tmlibs/cmn ...

**Comment by melekes** [comment]
Ok, so TestPEXReactorRunning is failing because addrbook#GetSelection is biased

```
D[05-24|15:09:59.251] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
D[05-24|15:09:59.391] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg=[pexRequest]
D[05-24|15:09:59.492] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
D[05-24|15:09:59.640] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg=[pexRequest]
D[05-24|15:09:59.640] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
D[05-24|15:09:59.888] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg=[pexRequest]
D[05-24|15:09:59.991] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
D[05-24|15:10:00.150] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg=[pexRequest]
D[05-24|15:10:00.251] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
D[05-24|15:10:00.388] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg=[pexRequest]
D[05-24|15:10:00.490] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
D[05-24|15:10:00.642] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg=[pexRequest]
D[05-24|15:10:00.749] Received message                             pex=1 src="Peer{MConn{127.0.0.1:41780} 5cf675ae3a82ed39f063cbb4ba411ce237c7140e in}" chId=0 msg="[pexAddrs [95f49b868fae0923f7d82487221c82ae8757ee3f@127.0.0.3:12156]]"
```

**Comment by melekes** [comment]
if I disable duplicate peer IP check, test passes

**Pull request merged**